src/server/replication_messages.rs

```diff
@@ -0,0 +1,249 @@
+            ReplicationChannel::Reliable,
+            Bytes::copy_from_slice(self.buffer.as_slice()),
+        );
+    }
+}
+
+/// A reusable message with component updates.
+///
+/// Contains last change tick, current tick and component updates since the last acknowledged tick for each entity.
+/// Requires init message with the same tick as last change tick to be applied to keep the world in valid state.
+/// The message will be manually split into packets up to max size that can be applied independently.
+/// Splits will happen per-entity to avoid weird behavior of partially changed entity.
+/// Sent over [`ReplicationChannel::Unreliable`] channel.
+///
+/// See also [Limits](../index.html#limits)
+#[derive(Deref, DerefMut, Default)]
+pub(super) struct UpdateMessage {
+    /// Entities and their data sizes.
+    entities: Vec<(Entity, usize)>,
+
+    /// Message data.
+    #[deref]
+    buffer: ReplicationBuffer,
+}
+
+impl UpdateMessage {
+    /// Clears the message.
+    ///
+    /// Keeps allocated capacity of the buffer.
+    fn reset(&mut self) -> bincode::Result<()> {
+        self.entities.clear();
+        self.buffer.reset();
+
+        Ok(())
+    }
+
+    /// Registers entity from buffer's entity data and its size for possible splitting.
+    pub(super) fn register_entity(&mut self) {
+        let data_size = self.buffer.as_slice().len() - self.buffer.entity_data_pos() as usize;
+        self.entities.push((self.buffer.data_entity(), data_size));
+    }
+
+    /// Splits message according to entities inside it and sends it to the specified client.
+    ///
+    /// Does nothing if there is no data to send.
+    fn send(
+        &mut self,
+        server: &mut RenetServer,
+        client_info: &mut ClientInfo,
+        last_change_tick: LastChangeTick,
+        replicon_tick: RepliconTick,
+        tick: Tick,
+    ) -> bincode::Result<()> {
+        let mut slice = self.buffer.as_slice();
+        if slice.is_empty() {
+            trace!("no updates to send for client {}", client_info.id);
+            return Ok(());
+        }
+
+        trace!("sending update message(s) to client {}", client_info.id);
+        const TICKS_SIZE: usize = 2 * mem::size_of::<RepliconTick>();
+        let mut header = [0; TICKS_SIZE + mem::size_of::<u16>()];
+        bincode::serialize_into(&mut header[..], &(*last_change_tick, replicon_tick))?;
+
+        let mut entities = Vec::new();
+        let mut message_size = 0;
+        for &(entity, data_size) in &self.entities {
+            const MAX_PACKET_SIZE: usize = 1200; // https://github.com/lucaspoffo/renet/blob/acee8b470e34c70d35700d96c00fb233d9cf6919/renet/src/packet.rs#L7
+            if message_size + data_size + header.len() > MAX_PACKET_SIZE {
+                let (message, remaining) = slice.split_at(message_size);
+                slice = remaining;
+                message_size = data_size;
+
+                let update_index = client_info.register_update(tick, entities.clone());
```

Comment:
I store regular ticks here just because I use `ClientInfo` when I iterate over the world and collect changes. This way I don't need to convert `RepliconTick` into `Tick` when I do it.